src/shims/unix/sync.rs

```diff
@@ -124,79 +124,40 @@ fn mutex_get_id<'tcx>(
+/// Returns the kind of a static initializer.
 fn kind_from_static_initializer<'tcx>(
     ecx: &MiriInterpCx<'tcx>,
-    contents: &[u8],
+    mutex_op: &OpTy<'tcx>,
 ) -> InterpResult<'tcx, MutexKind> {
-    let on_linux = &*ecx.tcx.sess.target.os == "linux";
-
-    let default_initializer =
-        ecx.eval_path(&["libc", "PTHREAD_MUTEX_INITIALIZER"]).assert_mem_place().ptr();
-    if contents == get_contents(ecx, default_initializer)? {
+    // Only linux has static initializers other than PTHREAD_MUTEX_DEFAULT.
+    if &*ecx.tcx.sess.target.os != "linux" {
         return Ok(MutexKind::new(ecx.eval_libc_i32("PTHREAD_MUTEX_DEFAULT")));
     }
 
-    if !on_linux {
-        throw_unsup_format!("unsuported pthread mutex initializer");
-    }
-
-    let recursive_initializer =
-        ecx.eval_path(&["libc", "PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP"]).assert_mem_place().ptr();
-    if contents == get_contents(ecx, recursive_initializer)? {
-        return Ok(MutexKind::new(ecx.eval_libc_i32("PTHREAD_MUTEX_RECURSIVE")));
-    }
-
-    let errorcheck_initializer = ecx
-        .eval_path(&["libc", "PTHREAD_ERRORCHECK_MUTEX_INITIALIZER_NP"])
-        .assert_mem_place()
-        .ptr();
-    if on_linux && contents == get_contents(ecx, errorcheck_initializer)? {
-        return Ok(MutexKind::new(ecx.eval_libc_i32("PTHREAD_MUTEX_ERRORCHECK")));
-    }
-
-    throw_unsup_format!("unsuported pthread mutex initializer");
-}
-
-/// Returns a byte slice holding the contents of the mutex pointed at by `ptr`.
-/// It will not fetch the initial "header" of the mutex, which is where miri stores the MutexId.
-/// This header is platform dependent, depending on the offset of the MutexId in `pthread_mutex_t`.
-fn get_contents<'tcx, 'ecx>(
-    ecx: &'ecx MiriInterpCx<'tcx>,
-    ptr: Pointer,
-) -> InterpResult<'tcx, &'ecx [u8]> {
-    // We want to check if the pthread_mutex_t pointed to by mutex_op is bitwise
-    // equal to a static initializer. However, we do not want to include the id (4 bytes)
-    // of the mutex in that check since mutex_get_or_create will have set it before
-    // calling this closure. We also want to discard the signature that macos stores
-    // in the first bytes.
-    let header_size = mutex_id_offset(ecx)?.strict_add(4);
-    let header_size = Size::from_bytes(header_size);
-
-    let ptr = ecx.ptr_offset_inbounds(
-        ptr,
-        header_size.bytes().try_into().expect("header len should fit in i64"),
-    )?;
-    ecx.read_bytes_ptr_strip_provenance(
-        ptr,
-        ecx.libc_ty_layout("pthread_mutex_t").layout.size() - header_size,
-    )
+    let offset = if ecx.pointer_size().bytes() == 8 { 16 } else { 12 };
+    let kind = ecx.deref_pointer_and_read(
+        mutex_op,
+        offset,
+        ecx.libc_ty_layout("pthread_mutex_t"),
+        ecx.machine.layouts.i32,
+    )?
+    .to_i32()?;
+    return Ok(MutexKind::new(kind));
```

Comment:
Please check that `kind` is a valid mutex kind.

In fact, probably it would make sense to make this an enum? Then all the other mutex functions won't have to compare this with libc constants any more...

Reply:
re making this an enum:

I don't disagree, but I am a bit unsure on how to do this. In my mind ideally the `MutexKind` would be a type known only to the pthreads implementation, but currently it lives with the generic mutex implementation. At least by making this an `i32` it would allow other mutex APIs other than pthreads to possibly reuse it. However if we make it an enum it might be too coupled with pthreads terminology.

As the code is now, we can't make the `AdditionalMutexData` generic so that each mutex implementation can define its own type to hold whatever data they want.

On the other hand this might be abstracting way to soon since there is no other implementation that needs additional data. What do you think?

Reply:
> As the code is now, we can't make the AdditionalMutexData generic so that each mutex implementation can define its own type to hold whatever data they want.

We could use `dyn Any` trickery.

But IMO it's not worth it. We can just say the enum lists all  mutex kinds of all implementations. :shrug: You can mark the enum `#[non_exhaustive]` to drive home that point, and force all matches to end with `_ => panic!("pthread mutex with invalid mutex kind")` or so.